### ISSUE TITLE
translated 404 page does not work

### ISSUE THREAD

**Issue by Gumichocopengin8** [withastro/starlight#1154]
### What version of `starlight` are you using?

0.13.0

### What version of `astro` are you using?

3.1.4 ~ 3.6.0

### What package manager are you using?

pnpm

### What operating system are you using?

macOS

### What browser are you using?

Safari

### Describe the Bug

I followed https://starlight.astro.build/guides/i18n/#translate-starlights-ui to setup i18n, but 404 page is not translated in Japanese somehow. The root language is `English`. whenever I access to `http://localhost:4321/ja/not-valid-page`, it shows English 404 page instead of Japanese one. I tried removing the custom `404.md` and creating the `i18` folder with `ja.json` inside, but still always shows the English 404 page. 
Is it a bug or am I missing something? 

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-yrwx8d?file=astro.config.mjs,src%2Fcontent%2Fi8n%2Fja.json

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by delucis** [comment]
Thank you for the issue @Gumichocopengin8!

Localized 404 pages are a bit tricky — we should probably document them in more detail. Notably, they will not work during `astro dev` but can be configured to work as expected in production on some hosts. Many hosts know how to use a root `404.html` file as the 404 page by default, but I haven’t found any that automatically understand custom 404 pages nested under language paths, so this always requires a bit of manual set-up.

1. Add a `404.md` for your chosen language, e.g. in your case `src/content/docs/ja/404.md`

2. Set up redirects on your host. This will vary a lot depending on your host. In my experience, Netlify does this best with their [Custom 404 page handling](https://docs.netlify.com/routing/redirects/redirect-options/#custom-404-page-handling). We have a [similar set-up for Vercel](https://github.com/withastro/starlight/blob/8d88850c64928878dc7d63c3ec90756690da8387/docs/vercel.json#L20) in the Starlight docs for this.

In `astro dev`, you can preview your 404 page by visiting it directly, e.g. <https://localhost:4321/ja/404>.

**Comment by Gumichocopengin8** [comment]
Thank you for your comment. with [vercel set-up](https://github.com/withastro/starlight/blob/8d88850c64928878dc7d63c3ec90756690da8387/docs/vercel.json#L20), it redirect to the language specific 404 page, so `404.text` key value in `i18n/ja.json` isn't effective with multiple-langue support?

**Comment by delucis** [comment]
> `404.text` key value in `i18n/ja.json` isn't effective with multiple-langue support?

Correct. Although, I think we _could_ generate translated 404 pages by default now I think about it. Would you expect a `ja/404` to be generated for you without having to add a `ja/404.md`? I think that should be possible just based on locale configuration.

**Comment by Gumichocopengin8** [comment]
> Would you expect a ja/404 to be generated for you without having to add a ja/404.md?

Yes. If `ja/404.md` or `i18/ja.json` is present, that overwrites the default value. However, `i18/ja.json` is only used when Japanese is the default language, right? so probably just `ja/404.md`??

**Comment by delucis** [comment]
> `i18/ja.json` is only used when Japanese is the default language, right?

Currently, yes. But we could in theory use it automatically for other languages too and only override that if a `404.md` for the language is found.

> I also expect no redirect settings for Vercel or Netlify

This will always be necessary due to how static sites work — Starlight has no control over 404 behaviour, so for multilingual 404s, there will always be some host-specific set-up unfortunately.

**Comment by Gumichocopengin8** [comment]
That makes sense. Thank you.

**Comment by Gumichocopengin8** [comment]
Sorry one more thing. 

Which elements are required for the 404 page? Is the following example sufficient or do I need other elements? 


```json
{
  "$schema": "https://openapi.vercel.sh/vercel.json",

  "routes": [
    { "src": "/(?<lang>[^/]*)/(.*)", "dest": "/$lang/404/", "status": 404 }
  ]
}
```


---
nit: Just realized that the `verbal.json` example is a bit outdated. I saw this warning. [routes](https://vercel.com/docs/projects/project-configuration#routes) should be replaced with something else. 

<img width="665" alt="image" src="https://github.com/withastro/starlight/assets/34010665/d34cb97d-7132-459a-871c-5c7b92c55d4d">

**Comment by delucis** [comment]
> Which elements are required for the 404 page? Is the following example sufficient or do I need other elements?

That example should work! `$schema` is not _required_ but gives you autocompletion and hints in your code editor, so is helpful.

> Also, I dont need to modify `src/content/config.ts` for multi-language 404 page, right?

No, you don’t need to.

> Just realized that the `verbal.json` example is a bit outdated. I saw this warning. [routes](https://vercel.com/docs/projects/project-configuration#routes) should be replaced with something else.

Correct. Unfortunately there is no way to do this kind of 404 redirect using Vercel’s newer configuration format. However, we were recommended to use `routes` by Vercel team members, so I think it’s safe to use.

**Comment by Gumichocopengin8** [comment]
Big thanks!! Everything is clear now!
